### DISCUSSION TITLE
How can I output Splink charts as .png for hard-copy outputs (e.g. pdf)

### DISCUSSION THREAD

**Discussion by RobinL** [moj-analytical-services/splink#926] [Q&A] [answered]
How can I output Splink charts as .png for hard-copy outputs (e.g. pdf)

**Comment by RobinL** [comment, marked as answer]
**Edit and Solution from maintainers 2023-06-16**
This question is now out of date due to [this PR](https://github.com/moj-analytical-services/splink/pull/1315)

All charts produced Splink are now full Altair Charts `alt.Chart()`.  You need to have Altair version 5 for them to work.

This means all charts can now be [more easily saved](https://altair-viz.github.io/user_guide/saving_charts.html) using e.g. `chart.save("myfile.png")`.  Note you can use file extensions `json`, `png`, `svg`, `pdf`, or `html` and others depending on what type output you want

Since all charts are now proper Altair charts, users experiencing difficulties should be asked to look at the [Altair documentation](https://altair-viz.github.io/user_guide/display_frontends.html).  


Previous answer:

Three options for doing this are contained in the following repo:
https://github.com/RobinL/splink_pdf_nbconvert_charts

Option 3 is probably the easiest, using the [vl-convert-python](https://pypi.org/project/vl-convert-python/0.4.0/) library

**Reply by aalexandersson** [reply]
Option 3 is very helpful, thanks! The only issue is that it requires static images, as is discussed [here](https://github.com/altair-viz/altair/pull/2701). For me as a Quarto user, it means two separate runs, where run 1 creates the .png file(s) and run 2 converts the .png file(s) to PDF.

A possibly better option 4 is to instead use `rsvg-convert` and .svg files, which I want to test once Quarto 1.3 prerelease becomes official:

Quarto version 1.3 is expected to be able to automatically convert .svg files (but not .png files) to PDF thanks to `rsvg-convert`, see  [https://quarto.org/docs/prerelease/1.3/pdf.html](https://quarto.org/docs/prerelease/1.3/pdf.html ).

**Reply by RobinL** [reply]
Looks like this is going to become even easier when altair 5 is released:
https://twitter.com/jonmmease/status/1631278647939137536

**Reply by aalexandersson** [reply]
I am using splink 3.9.2 with altair 5.0.1 (on Windows with Python 3.11) but I get the error 
`NameError: name 'chart' is not defined` when I simply use `chart.save("myfile.png")` as suggested on 6/16.

What is an example of the correct syntax to use and save a missingness chart of a dataframe to PDF? The code below fails with the error just mentioned.
```
import altair as alt  
linker = DuckDBLinker([df_orig, df_noise], settings,
    input_table_aliases=["no_noise", "with_noise"])                           
c_noise = linker.missingness_chart()             
chart.save("c_noise.png")
```

This is the traceback:
```
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Input In [5], in <cell line: 0>()
      6 import altair as alt                             # 
      7 c_noise = linker.missingness_chart()             # 
----> 8 chart.save("c_noise.png")

NameError: name 'chart' is not defined
NameError: name 'chart' is not defined
```

**Reply by RobinL** [reply]
It should be:

```
c_noise = linker.missingness_chart()             
c_noise.save("c_noise.png")
```

**Reply by aalexandersson** [reply]
Yes, sorry about that (https://altair-viz.github.io/user_guide/saving_charts.html) but it gives another error message:


`Unsupported Vega-Lite version string v5_8`

```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "F:\Monograph_2023\.venv\Lib\site-packages\altair\vegalite\v5\api.py", line 948, in save
    result = save(**kwds)
             ^^^^^^^^^^^^
  File "F:\Monograph_2023\.venv\Lib\site-packages\altair\utils\save.py", line 154, in save
    mimebundle = spec_to_mimebundle(
                 ^^^^^^^^^^^^^^^^^^^
  File "F:\Monograph_2023\.venv\Lib\site-packages\altair\utils\mimebundle.py", line 51, in spec_to_mimebundle
    return _spec_to_mimebundle_with_engine(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "F:\Monograph_2023\.venv\Lib\site-packages\altair\utils\mimebundle.py", line 111, in _spec_to_mimebundle_with_engine
    png = vlc.vegalite_to_png(
          ^^^^^^^^^^^^^^^^^^^^
RuntimeError: Unsupported Vega-Lite version string v5_8
```

**Reply by aalexandersson** [reply]
It seems that the offending package mimebundle.py requires version 5.2 of vlconvert:
```
    if normalized_engine == "vlconvert":
        import vl_convert as vlc
        from ..vegalite import SCHEMA_VERSION

        # Compute VlConvert's vl_version string (of the form 'v5_2')
        # from SCHEMA_VERSION (of the form 'v5.2.0')
        vl_version = "_".join(SCHEMA_VERSION.split(".")[:2])
        if format == "vega":
            vg = vlc.vegalite_to_vega(spec, vl_version=vl_version)
            return {"application/vnd.vega.v5+json": vg}
        elif format == "svg":
            svg = vlc.vegalite_to_svg(spec, vl_version=vl_version)
            return {"image/svg+xml": svg}
        elif format == "png":
            png = vlc.vegalite_to_png(
                spec,
                vl_version=vl_version,
                scale=kwargs.get("scale_factor", 1.0),
            )
            return {"image/png": png}
        else:
```

But the command `pip list` states that I use
`vl-convert-python    0.5.0`
which is less than 5.2. According to https://pypi.org/project/vl-convert-python/#history
version 0.5.0 is from Oct 29, 2022 and the current version is 0.10.3 from June 17. Maybe I need to upgrade vl-convert-python?

**Reply by RobinL** [reply]
That sounds right. When you install Altair 5.0.1 it doesn't install vl-convert-python. Instead it shows a message telling you to install it when you try and save out to PNG. It sounds like they hadn't considered the case where it is installed, but just an old version.  Well done for figuring it out (assuming you have!), I was a bit stumped

**Reply by aalexandersson** [reply]
Thank you. The only graph that still fails for me is the waterfall chart because your previous fix at [discussion 1044](https://github.com/moj-analytical-services/splink/discussions/1044) no longer works in Altair 5.0.1. Any ideas?

```
spec = chart.spec
```
results in this error:
`AttributeError: 'LayerChart' object has no attribute 'spec'`

If I instead simply try the new Altair 5 standard `chart.save() `which now works for all my other charts:
```
c_waterfall = linker.waterfall_chart(records_to_plot, filter_nulls=False)
c_waterfall.save("c_waterfall.png")
```
Then I get the error message in mimetype.py at line 111 as in discussion 1044:
```
ValueError: Vega to SVG conversion failed:
TypeError: Cannot read properties of undefined (reading 'marktype')
```

If I try to use the same fix as in discussion 1044:
```
c_waterfall = linker.waterfall_chart(records_to_plot, filter_nulls=False)
del c_waterfall["layer"][0]["layer"][1]["encoding"]["x"]["axis"]["labelExpr"]
c_waterfall.save("c_waterfall.png")
```
then I get same error:
```
ValueError: Vega to SVG conversion failed:
TypeError: Cannot read properties of undefined (reading 'marktype')
```

**Reply by RobinL** [reply]
Yeah, looks like this bug is still not fixed in `vl-convert-python`.  I should probably report it/try to fix it but need to spend some time to come up with a simpler reprex

The equivalent workaround in Altair is:
```
chart = linker.waterfall_chart(records_to_plot, filter_nulls=False)
chart.layer[0].layer[1].encoding.x.axis.labelExpr = ''
```
And then e.g.
```
chart.save("my_chart.png", scale_factor=2)
```

(Actually strictly speaking the exact equivalent is 
```
from altair.utils.schemapi import UndefinedType
chart.layer[0].layer[1].encoding.x.axis.labelExpr = UndefinedType()
```
but they have the same effect)

Thanks for flagging this.  I've updated #1044 to include this new solution

**Reply by aalexandersson** [reply]
The modified fix worked for me. Splink with Altair 5 is much more user-friendly for graphics. Thanks again for all your help!

**Reply by RobinL** [reply]
> The modified fix worked for me. Splink with Altair 5 is much more user-friendly for graphics. Thanks again for all your help!  

You're very welcome.  Glad to hear it and sorry for the teething troubles!

**Reply by RobinL** [reply]
Just to note I've raised [an issue](https://github.com/vega/vl-convert/issues/72) against `vl-convert-python` so hopefully this will be fixed (or they'll let me know if we're doing something wrong in Splink)

**Reply by RobinL** [reply]
@aalexandersson this is now [fixed](https://github.com/vega/vl-convert/issues/72) in the latest version of `vl-convert-python`, so if you `pip install vl-convert-python==0.11.1` you no longer need the workaround on the waterfall chart.  thanks for your patience with this issue!

**Reply by aalexandersson** [reply]
Thank you. But commenting out the fix using `vl-convert-python==0.11.1` fails for me:

`chart.layer[0].layer[1].encoding.x.axis.labelExpr = ''`

Error as before:
```
ValueError: Vega to SVG conversion failed:
TypeError: Cannot read properties of undefined (reading 'marktype')
```

**Reply by RobinL** [reply]
Did you definitely restart the kernel/Python session after installing 0.11.1?  In tests on my machine, i no longer get the error (i tried in python 3.8 and 3.11).

If you have a moment, would you mind trying out the small reproducible example in the issue [here](https://github.com/vega/vl-convert/issues/72).  Does that also fail for you?  If so might be worth putting a message on that thread...

**Reply by aalexandersson** [reply]
Sorry, the fix works for me too now. I am working in a venv. I mistakenly forgot to update within the venv (or to not use a venv). Thanks again for the help!
